### PR TITLE
Allow to feed a value in another query's cache and remove `WithOptConstParam`

### PR DESCRIPTION
**Pull request by cjgillot**
I used it to remove `WithOptConstParam` queries, as an example.

The idea is that a query (here `typeck(function)`) can write into another query's cache (here `type_of(anon const)`). The dependency node for `type_of` would depend on all the current dependencies of `typeck`.

There is still an issue with cycles: if `type_of(anon const)` is accessed before `typeck(function)`, we will still have the usual cycle.  The way around this issue is to `ensure` that `typeck(function)` is called before accessing `type_of(anon const)`.

When replayed, we may the following cases:
- `typeck` is green, in that case `type_of` is green too, and all is right;
- `type_of` is green, `typeck` may still be marked as red (it depends on strictly more things than `type_of`) -> we verify that the saved value and the re-computed value of `type_of` have the same hash;
- `type_of` is red, then `typeck` is red -> it's the caller responsibility to ensure `typeck` is recomputed *before* `type_of`.

As `anon consts` have their own `DefPathData`, it's not possible to have the def-id of the anon-const point to something outside the original function, but the general case may have to be resolved before using this device more broadly.

There is an open question about loading from the on-disk cache.  If `typeck` is loaded from the on-disk cache, the side-effect does not happen. The regular `type_of` implementation can go and fetch the correct value from the decoded `typeck` results, and the dep-graph will check that the hashes match, but I'm not sure we want to rely on this behaviour.

I specifically allowed to feed the value to `type_of` from inside a call to `type_of`.  In that case, the dep-graph will check that the fingerprints of both values match.

This implementation is still very sensitive to cycles, and requires that we call `typeck(function)` before `typeck(anon const)`.  The reason is that `typeck(anon const)` calls `type_of(anon const)`, which calls `typeck(function)`, which feeds `type_of(anon const)`, and needs to build the MIR so needs `typeck(anon const)`.  The latter call would not cycle, since `type_of(anon const)` has been set, but I'd rather not remove the cycle check.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/query.rs [hunk 1]
```diff
@@ -231,6 +254,52 @@ macro_rules! define_callbacks {
             })*
         }
 
+        impl<$tcx> TyCtxtFeed<$tcx> {
+            $($(#[$attr])*
+            #[inline(always)]
+            pub fn $name<V>(
```
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/query.rs:260]: Would it make sense to only allow this for queries which explicitly opt-in? And then maybe not allow setting any provider for such queries. I think that makes the information flow a lot clearer.

File: compiler/rustc_query_system/src/dep_graph/graph.rs [hunk 2]
```diff
@@ -489,6 +489,87 @@ impl<K: DepKind> DepGraph<K> {
...
+    pub fn with_feed_task<Ctxt: DepContext<DepKind = K>, A: Debug, R: Debug>(
+        &self,
+        node: DepNode<K>,
+        cx: Ctxt,
+        key: A,
+        result: &R,
+        hash_result: fn(&mut StableHashingContext<'_>, &R) -> Fingerprint,
+    ) -> DepNodeIndex {
+        if let Some(data) = self.data.as_ref() {
+            if let Some(dep_node_index) = self.dep_node_index_of_opt(&node) {
```
Comment (cjgillot) [inline compiler/rustc_query_system/src/dep_graph/graph.rs:516]: FIXME: explain what happens in that case.

**Review by cjgillot — COMMENTED** [review]
Thanks rustbot. So much for the discretion of a WIP PR :smile:

**Comment by cjgillot** [comment]
r? compiler

**Comment by lcnr** [comment]
looked through this PR and I am really happy about not seeing `WithOptConstParam` anymore :sparkles:

The query system is complex enough for me to not be comfortable approving this. My thoughts:
- we should probably change the `type_of` query to ICE if `opt_const_param_of` returns `Some`. Or actually, we should remove the `opt_const_param_of` query entirely and put the branches from there into `type_of` and ICE for branches which would result in `Some`.
- doing that should change the buggy behavior when not using `feed` from query cycle to ICE (with a good message). So I am not that concerned about the impact on other rustc devs using `type_of`, especially compared to understanding `WithOptConstParam` which is a lot more involved.
- being feedable should be opt-in, as mentioned by @bjorn3. We should probably even change it so that queries are either feedable or provided. For that we would have to split `type_of` into `type_of` and `anon_const_type` or whatever :thinking: 
- if we have `type_of` calling another query which then feeds the value of `type_of` (which is currently getting computed), we should ICE I think? I don't think this is currently covered, but haven't looked into the query impl too much ^^

going to hand off this review to incr comp people, but I am looking forward to this PR landing. I do think it requires an MCP though.

r? @wesleywiser maybe, @michaelwoerister isn't available much rn afaict

**Comment by cjgillot** [comment]
>     * we should probably change the `type_of` query to ICE if `opt_const_param_of` returns `Some`. Or actually, we should remove the `opt_const_param_of` query entirely and put the branches from there into `type_of` and ICE for branches which would result in `Some`.
> 

This is not really possible yet.  I could remove the case where `tcx.typeck` was invoked, but other uses remain.  For instance, types in signatures cannot be dealt with that way, because `const_evaluatable_predicates_of` needs them.

>     * doing that should change the buggy behavior when not using `feed` from query cycle to ICE (with a good message). So I am not that concerned about the impact on other rustc devs using `type_of`, especially compared to understanding `WithOptConstParam` which is a lot more involved.
> 

Mmh... I wonder where to put this message.  Do you have an example of such misuse?

>     * being feedable should be opt-in, as mentioned by @bjorn3. We should probably even change it so that queries are either feedable or provided. For that we would have to split `type_of` into `type_of` and `anon_const_type` or whatever thinking
> 

Done.

>     * if we have `type_of` calling another query which then feeds the value of `type_of` (which is currently getting computed), we should ICE I think? I don't think this is currently covered, but haven't looked into the query impl too much ^^
> 

We do not ICE, we check that both values match.

**Comment by camsteffen** [comment]
I based #100720 on this PR and I think it shows an interesting use case for query "feeding".

> We should probably even change it so that queries are either feedable or provided.

I found it useful to have a query that is both provided and feedable. Though it is only fed by itself, recursively.

**Comment by nikomatsakis** [comment]
I left a [comment on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/241847-t-compiler.2Fwg-incr-comp/topic/Setting.20outputs.20from.20queries/near/299224853) but I wanted to raise the concern on the PR as well--

I think we need to limit is so that you cannot "feed" a value from one query to another *unless* the first query created the key used by the second query. Otherwise, you run the very real risk of non-deterministic results if the query order changes. 

The danger is something like this:

* Query A(X) sets the result of B(X).
* Query C(X) reads B(X).
* Query D(X) first executes A(X) and then C(X).

Now if you start by invoking `C`, you get one value (whatever `B(X)` computes).

But if you start by invoking `D`, you get a different value (whatever `A(X)` stored).

If I understand the current code, it works by setting the "inputs" for B(X) to be the inputs that A(X) has observed thus far, but that's not enough to prevent the above problem.

**Comment by oli-obk** [comment]
For some of these cases (like the one done in this PR) this is not an issue, as `B(X)` has no impl and will either invoke `A(X)` to get its value or will ICE. But I agree that it's fragile to rely on this.

One way I have considered for making this robust is to have [TyCtxt::create_def](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.create_def) return a `NewDefId` wrapper type that can only created by `create_def`. Then `tcx.feed().query_name` would only accept arguments of type `NewDefId` where they currently required `DefId`. You can always `NewDefId::into_inner` to use it normally, but for feeding you'll need the `NewDefId`. Then we just need to make sure no one returns `NewDefId` from a query, but that seems doable (just implement no traits at all for it).

for this PR that would mean to stop generating `DefId`s for `AnonConst` in AST->HIR lowering and instead leave that to typeck of the owner of the anon const

**Comment by cjgillot** [comment]
> I left a [comment on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/241847-t-compiler.2Fwg-incr-comp/topic/Setting.20outputs.20from.20queries/near/299224853) but I wanted to raise the concern on the PR as well--
> 

That's the point of this PR: listing risks and answering to them.

> I think we need to limit is so that you cannot "feed" a value from one query to another _unless_ the first query created the key used by the second query. Otherwise, you run the very real risk of non-deterministic results if the query order changes.
> 

I need to think a bit more what this would imply on the code structure around `AnonConst` and metadata emission.

> The danger is something like this:
> 
>     * Query A(X) sets the result of B(X).
> 
>     * Query C(X) reads B(X).
> 
>     * Query D(X) first executes A(X) and then C(X).
> 
> Now if you start by invoking `C`, you get one value (whatever `B(X)` computes).
> 
> But if you start by invoking `D`, you get a different value (whatever `A(X)` stored).
> 
> If I understand the current PR description, it works by setting the "inputs" for B(X) to be the inputs that A(X) has observed thus far, but that's not enough to prevent the above problem.

I cannot avoid this in general.  However, I have several mitigations:
- if `A(X)` feeds a value first, we never execute `B(X)`, so nothing to do;
- if `B(X)` is marked green and not computed, and then `A(X)` feeds a value, we assert that the hashes match;
- if `B(X)` is computed first, and then `A(X)` feeds a value, we assert that the values match;
- if `B(X)` calls `A(X)` which feeds a value, we assert that what `A(X)` put and what `B(X)` computes have the same hash (in that case, I recommend that `B(X)` tries to return `B(X)` by calling itself).

I think that these are all the cases that could happen.  Do you see another?

**Comment by cjgillot** [comment]
> For some of these cases (like the one done in this PR) this is not an issue, as `B(X)` has no impl and will either invoke `A(X)` to get its value or will ICE. But I agree that it's fragile to rely on this.
> 
> One way I have considered for making this robust is to have [TyCtxt::create_def](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.create_def) return a `NewDefId` wrapper type that can only created by `create_def`. Then `tcx.feed().query_name` would only accept arguments of type `NewDefId` where they currently required `DefId`. You can always `NewDefId::into_inner` to use it normally, but for feeding you'll need the `NewDefId`. Then we just need to make sure no one returns `NewDefId` from a query, but that seems doable (just implement no traits at all for it).
> 
> for this PR that would mean to stop generating `DefId`s for `AnonConst` in AST->HIR lowering and instead leave that to typeck of the owner of the anon const

I just tried this.  Creating the `DefId`s for `AnonConst` during typeck is be fine, it can even be bundled into `ty::Const::from_anon_const`.

However, this breaks one very used invariant of HIR: that a `Body` has a `body_owner_def_id`.  Creating the `DefId` for `AnonConst` later means that the HIR cannot know the relationship between the `AnonConst`'s node `HirId`, and the created `DefId`.  As the `tcx.hir().local_def_id` map becomes incomplete, code that relies on the ability to go from one to the other (like typechecking the AnonConst, building MIR for it) needs a refactor to avoid using that map.

I've started that refactor on a separate branch, but it's larger than I anticipated.

**Comment by wesleywiser** [comment]
Hi @cjgillot, was this PR superseded by #104940 and can now be closed or are there other changes here you'd like to merge after rebasing?

**Comment by nbdd0121** [comment]
That PR just include the infra for query feeding, but doesn't include the `WithOptConstParam` removal bit.

**Comment by cjgillot** [comment]
Update: this is blocked on a larger refactor. During the meeting about this PR, it was decided that we should restrict `feed` to newly created `LocalDefId`s. In order to do that, we need to (2) delay the creation of `LocalDefId` for `AnonConst` to the first use, (1) remove the `HirId -> LocalDefId` map which becomes a footgun.

Step (1) is waiting for https://github.com/rust-lang/rust/pull/103902 to be merged as a prerequisite.

**Comment by Zoxc** [comment]
I've not been able to find any problems with the concept of feeding queries yet. To get the current state to work better we should let queries define exactly one query instance which could feed them. For example:
```rust
query type_of(key: DefId) -> Ty<'tcx> {
    feedable {
        if is_anon_const(key) {
            Some(query::type_of(parent_of(key)))
        } else {
            None
        }
    }
}
```
When executing `type_of` we check if we need to run another query to be fed or if we just execute our provider. When a query is being fed we check if we are the query that is supposed to feed it.

I also prefer a `HashMap::insert` like syntax for feeding, like `tcx.feed().type_of(key, value)`.

**Comment by oli-obk** [comment]
> When executing `type_of` we check if we need to run another query to be fed or if we just execute our provider. When a query is being fed we check if we are the query that is supposed to feed it.

What would be the advantage of that design?

**Comment by Zoxc** [comment]
It's not sensitive to query execution order as query that can be fed will always use the value from the query that feeds it.

You can also express the scheme where the key stores a "creating query" with `feedable { Some(key.query) }`. Most likely that information could be encoded more efficiently if needed though.

**Comment by oli-obk** [comment]
That seems like a future extension as it would be more general and allow new things. We should probably first come up with a use case for it.

The current scheme is a bit simpler, even if more limited.

Even if not simpler from an impl perspective, it is simpler for rustc devs I'd argue.

**Comment by Zoxc** [comment]
One thing I though of is that it could be used an alternative solution to generate MIR for generators during typeck as we could feed the types required for that to another query. Not sure if that's better than the current approach. It does seem like it could be helpful other situations where we run into query cycles due to imprecise dependency tracking.

**Comment by cjgillot** [comment]
The latest push keeps the same design as the original. I still haven't found a design to create the `AnonConst`'s `LocalDefId` just before feeding the type.

> One thing I though of is that it could be used an alternative solution to generate MIR for generators during typeck as we could feed the types required for that to another query. Not sure if that's better than the current approach. It does seem like it could be helpful other situations where we run into query cycles due to imprecise dependency tracking.

I started an experiment in https://github.com/rust-lang/rust/pull/107406 for generators.

File: compiler/rustc_middle/src/ty/context.rs [hunk 3]
```diff
@@ -447,6 +447,9 @@ impl<'tcx> TyCtxt<'tcx> {
     pub fn feed_local_crate(self) -> TyCtxtFeed<'tcx, CrateNum> {
         TyCtxtFeed { tcx: self, key: LOCAL_CRATE }
     }
+    pub fn feed_def_id(self, key: LocalDefId) -> TyCtxtFeed<'tcx, LocalDefId> {
```
Comment (oli-obk) [inline compiler/rustc_middle/src/ty/context.rs:447]: We could special case this to be a `feed_type_of_anon_const`, ensured by assertions to only be used for the specific cases we want.

I'll write up a hackmd with a plan to use the regular feeding after solving the issues around iterating over body owners if said iteration or later logic may create more body owners

**Comment by oli-obk** [comment]
Do we already have incremental tests for situations where the type of the const changes?

r=me with those added or verified we have them

**Comment by cjgillot** [comment]
I added a revert of #108046 as last commit, as this corner case is not impossible any more.

There are 2 incremental tests. Do you have other ideas for test cases?

**Comment by apiraino** [comment]
Visiting this PR, checking for progress. @cjgillot what's the actual status? Our understanding on the [T-compiler meeting on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202022-12-01/near/313304319) was that is was complemented by #104940.

What's next for this, is it waiting on review? thanks

File: compiler/rustc_query_system/src/query/plumbing.rs [hunk 4]
```diff
@@ -410,9 +410,26 @@ where
         // as its feeding query had. So if the fed query is red, so is its feeder, which will
         // get evaluated first, and re-feed the query.
         if let Some((cached_result, _)) = cache.lookup(&key) {
-            panic!(
-                "fed query later has its value computed. The already cached value: {}",
-                (query.format_value())(&cached_result)
+            let Some(hasher) = query.hash_result() else {
+                panic!(
+                    "fed query later has its value computed. The already cached value: {}",
```
Comment (oli-obk) [inline compiler/rustc_query_system/src/query/plumbing.rs:415]: maybe note that this works for crates without `no_hash`

**Comment by oli-obk** [comment]
The PR description is not up to date anymore. There's no `anon_const_type_of`, but instead the `type_of` query is fed directly.

**Comment by oli-obk** [comment]
r=me with description updated (afaict it still matches the impl, if one replaces `anon_const_type_of` with `type_of` everywhere).

**Comment by nbdd0121** [comment]
https://github.com/rust-lang/rust/blob/1f5768bc67ecb025342770e14e03699699965706/compiler/rustc_middle/src/ty/consts.rs#L129

One reference to `WithOptConstParam` remaining in the comment

**Pull request merged**